gpbft/participant.go

```diff
@@ -91,31 +94,38 @@ func (p *Participant) ValidateMessage(msg *GMessage) (checked bool, err error) {
 }
 
 // Receives a Granite message from some other participant.
-// The message is delivered to the Granite instance if it is for the current instance,
-// else it is dropped.
-// This method *does not check message validity*.
-// The message must have been previously validated with ValidateMessage indicating success.
-// Since messages for future instances cannot be validated, a valid message
-// can only be for the current or some previous instance (hence dropping if not current).
-// Returns whether the message was accepted for the instance, and an error if it could not be
-// processed.
-func (p *Participant) ReceiveMessage(msg *GMessage) (accepted bool, err error) {
+// If the message is for the current grant instance, it is delivered immediately.
+// If it is for a future instance, it is locally queued, to be delivered when that instance begins.
+// If it is for a previous instance, it is dropped.
+//
+// This method checks message validity only if the validated parameter is false.
+// Validity can only be checked when the messages instance begins.
+// Returns whether the message was for the current instance, and an error if it was invalid or
+// could not be processed.
+func (p *Participant) ReceiveMessage(msg *GMessage, validated bool) (accepted bool, err error) {
```

Comment:
Should we include sanity checks that we can include for future instances directly here? (Mitigate DOS attacks). I.e.: a strong quorum of decide must contain valid signatures (even if we cannot really verify that the signatures constitute a strong quorum yet)

Reply:
We can't even verify the signatures yet, because we don't know the keys to use. We could probably use a heuristic but I would prefer to defer thinking hard about that so we can get the basic thing working sooner